### PR TITLE
Add ability to view new add-ons

### LINKED ISSUE #16681: Add ability to view new available add-ons
### Is your feature request related to a problem? Please describe.
Users may want to be notified when new add-ons are available from the store.
### Describe the solution you'd like
- A New add-ons tab maybe added to the store, where last compatible add-ons not cached would be presented.
- A dialog or other means, like an ui.message or a toast notification, would notify users when new add-ons are available, for example at startup or at scheduled intervals (each hour may be reasonable.
- The dialog will be the store focusing the New add-ons tab, where the description and other details, an an action to install add-ons, would be available.
### Describe alternatives you've considered
An add-on maybe created to notify about new add-ons, but it maybe better to include this in the core adding a new tab to the store.
### Additional context
A bot used on Mastodon, on a Feeds subgroup of the add-ons mailing list and on a spanish mailing list is used by some people to be notified about new add-ons:

- [Bot repo](https://github.com/nvdaes/nvdaStoreMastodonBot)
- [Bot profile on Mastodon](https://mastodon.chrisduffley.com/@nvdastore)

### PR DESCRIPTION
**Pull request by nvdaes**
- Add support to show new add-ons in the store
- Add schedule job to show new add-ons at startup
- Check for new add-ons before running scheduled job to show them
- Fix lint

### Link to issue number:
Closes #16681

### Summary of the issue:

List of available add-ons is very long, and some users may wish to show a list of recent add-ons with their description and other details to decide if they wish to install them.

### Description of user facing changes
- A new combo box to filter add-ons by release date has been added to the available add-ons tab of the store.
- A new combo box to determine the period of time to consider add-ons as new has been added to the add-on store settings panel.

### Description of development approach
- Cached compatible add-ons are saved in a backup file on NVDA exit, or at the datetime determined in the add-on store settings, for example, at startup, weekly or monthly, to be calculated considering the last modification date of the file used to cache old add-ons.
- New add-ons can be retrieved comparing the current compatible available add-ons with previous add-ons. Specifically, if an add-on with a non existing id or a new version of an existing add-on is available, it will be listed in the New add-ons option of the Release date filter, unles it's already installed.
### Testing strategy:
Tested manually:
- Removed the `_cachedCompatibleAddons-old.json` file, saved when NVDA is started, and check that the second option of the Release date filter combo box shows that NVDA needs to be restarted to show new add-ons
- Removed an entry corresponding to an old add-on in the json file, and check that it's listed as a new add-on.
- Change the version number of an old add-on, and check that the add-on with the current available version is listed as a new add-on.
- Check that, if an add-on is installed, it's not listed as a new add-on.
- Check that dates to consider add-ons as new (recent) are properly updated in the Release date filter combo box, after changing the corresponding option in the add-on store settings panel.

### Known issues with pull request:
None
### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Introduced the ability to filter add-ons by release date in the add-on store.
  - Added a "Reset New Add-ons" feature, allowing users to specify reset periods (Monthly, Weekly, At startup).
  - Implemented a new status `NEW` for recently published add-ons, enhancing add-on discovery and management.

- **Enhancements**
  - Updated add-on store settings panel with options for resetting the new add-ons status and filtering by release date.
  - Added new controls and logic in the add-on store UI for better management of add-on statuses and filters.

- **Documentation**
  - Updated user guide and changelog to reflect new add-on management features and settings.

### CODE REVIEW

**Comment by seanbudd** [comment]
@coderabbitai review

**Comment by seanbudd** [comment]
as requested in #16681, can you seek UX feedback before marking this as ready?

**Comment by nvdaes** [comment]
Sean wrote: 

> as requested in #16681, can you seek UX feedback before marking this as ready?

I'm seeking for feedback:

- [On Mastodon](https://mastodon.chrisduffley.com/@nvdastore/112659644984939911)
- [On the add-ons mailing list](https://nvda-addons.groups.io/g/nvda-addons/message/24704)
- [On a spanish mailing list](https://nvdaes.groups.io/g/lista/topic/solicitud_para_mostrar_nuevos/106829024)
- Commenting this to a friend

A few people of the spanish mailing list (two persons) would like to see this feature, and my friend said me that it's a good suggestion, but I didn't get more precise feedback.
I'd try to share an artifact so people can test this in a portable copy of NVDA.

**Comment by seanbudd** [comment]
Thanks, we really need validation that this is the right approach and useful to users

**Comment by Qchristensen** [comment]
A different approach to this PR, but rather than have a new tab in the add-on dialog, would it be a better experience for users to have a "Last updated" date in the table list of add-ons, and have the ability to sort by various table columns.  I would argue the date an add-on was last updated is more meaningful than the "add-on version" we currently list, which is only really relevant to THAT add-on, as every add-on uses its own version format.

If really wanted, you could then flag add-ons added since the add-on store was last opened as "new" (or have a flag the user could reset on demand).

**Comment by nvdaes** [comment]
Quentin wrote:

> A different approach to this PR, but rather than have a new tab in the add-on dialog, would it be a better experience for users to have a "Last updated" date in the table list of add-ons...

But a some add-ons may be very old in terms of the last update date but they may work perfectly, and instead, newer add-ons may need to be more maintained due to user requests or due of components used. Anyway this is a separate issue and needs to be triaged.
 See #15278 



> the ability to sort by various table columns. I would argue the date an add-on was last updated is

This is a separate issue

See #15277

**Comment by Qchristensen** [comment]
> Quentin wrote:
> 
> But a some add-ons may be very old in terms of the last update date but they may work perfectly, and instead, newer add-ons may need to be more maintained due to user requests or due of components used. Anyway this is a separate issue and needs to be triaged. See #15278

True, but in the case of an add-on whose last version is three years old but still works great.... the proposed new add-ons info won't be of help either :)

> > the ability to sort by various table columns. I would argue the date an add-on was last updated is
> 
> This is a separate issue See #15277

My suggestion was that sorting the existing available add-ons list by latest update meets the same goal (of seeing which add-ons are new) as the proposed "new available add-ons" tab

**Comment by nvdaes** [comment]
> True, but in the case of an add-on whose last version is three years old but still works great.... the proposed new add-ons info won't be of help either :)

The goald of this PR is to see just recent add-ons to decide if they should be installed, seeing the description and details. The last date proposal will be useful as well, similar to this PR, but imo it would require to work in three repos: addon-datastore, addon-datastore-validation and NVDA with GitHub Actions, and may take a lot of time and work, and furthermore it's not triaged.

**Comment by nvdaes** [comment]
Here's [message 5002 of a spanish mailing list](https://nvdaes.groups.io/g/lista/message/5002), where someone has tested the add-on.
I provided a json file to simulate the presence of new add-ons. This tester says that he likes the idea and its application. He finds practical to view info about new add-ons separately.
I commented that the idea of having info  about dates seems good to me, though perhaps I prefer my approach since having too many info in the table (in columns) is more difficult (slow) to read.
About the date, since it wasn't triaged, I think that the only option was to work on this approach.

**Comment by nvdaes** [comment]
As requested by @Christianlm, another tester of this feature, I've created a combobox in the add-on store settings panel to reset new add-ons at startup, weekly or monthly. Also, the Notify when new add-ons are available is disabled by default,  since, if new add-ons are reset monthly (as set by default now), having the notification option enabled may result in seeing the same add-ons at startup.
System tests have failures. I'll test this more. Also, I'm thinking in trying to add the starting and end dates used to consider add-ons as new, for example, from 1 July 2024 to 31 July 2024, where add-ons would be reset.
See [Chris message requesting options to reset add-ons monthly or weekly](https://nvdaes.groups.io/g/lista/message/5011).

**Comment by nvdaes** [comment]
@coderabbitai review

**Comment by Adriani90** [comment]
I am actually not really in favor of a new tab. I vote rather for a sorting combo box in the available add-ons tab.
sort by combo box:
* Option 1: by name
* Option 2: By date of last update
* Option 3: by date of Submission approval

The current UX of this PR is too complex in my view.
If users can sort the list of available add-ons as suggested above, there is no notification needed.

**Comment by nvdaes** [comment]
@Adriani90 ,  I think that sorting by submission approval of an add-on id is not related to this feature. Sorting by date was not triaged. See a previous comment where this was explained.

**Comment by Adriani90** [comment]
But submission approval means new add-on. Or am I wrong?Von meinem iPhone gesendetAm 01.07.2024 um 18:13 schrieb Noelia Ruiz Martínez ***@***.***>:﻿
@Adriani90 ,  I think that sorting by submission approval of an add-on id is not related to this feature. Sorting by date was not triaged. See a previous comment where this was explained.

**Comment by nvdaes** [comment]
This has mixed feedback, but since add-ons date is not shown in the store, and I don't know if dates should be presented in the add-ons list or in the details panal (I probably would prefer this last option), I mark this as ready for review.

**Comment by CyrilleB79** [comment]
I have not yet been able to review or test this design.
But reading the description and the comments of this PR, I do not feel it's worth to complexify the GUI for such information.
That's a provisional feedback though and I may change my mind after a more careful review and test of this PR if/when I can.
If you do not wish to wait anymore and since I am not sure to have time to provide more information, feel free to ignore my feedback.

**Comment by nvdaes** [comment]
Cyrille wrote:

> If you do not wish to wait anymore and since I am not sure to have time to provide more information, feel free to ignore my feedback.

I think we can wait, this is not something that requires to act urgently.
I think that having info about add-ons added in a certain period of time (recent add-ons) can be useful, since older add-ons may have been seen and this may be useful to help to decide if an add-on is installed.
We had the website with feeds, though new versions weren't shown as new pages. We have a bot followed by some users, and I think that having this inside NVDA would be interesting.
About complexing the gui, I think that if we want this feature this is needed. Imo a new tab is the less complex option since tabs can be ignored.
But other designs are possible, of course. Thanks for your feedback.

**Comment by XLTechie** [comment]
Thanks for your work on this, @nvdaes.

Personally though, I would prefer to see something closer to what @Adriani90 suggested.

Or maybe a combobox for sorting, and another for filtering, while in Available add-ons tab.

Sorting: name (A-Z), name (Z-A), updated date, added date.

Filter: new in last week, new in last month, new this year, updated in last week, updated in last month, updated this year, all.

Where last week and last month, are today's date minus seven days or 31 days, not necessarily this calendar week/month which might have just started a minute ago.

Regardless of this idea or another, I think we should think carefully about making the interface more complex. Users already find it to be complicated as it is.

**Comment by nvdaes** [comment]
Thanks for your feedback, @XLTechie.
Since here  you all prefer a combo box, I'll try to replace the tab with a combo box, filtering new add-ons but using the same procedure of caching add-ons in a json file, since, as you know, add-on metadata doesn't include a timestamp and an  issue related to dates, as explained before, is not triaged, so the only possibility is to work without having dates.
I'll work in the combobox the next days and we can always roll back to the tab design.

File: source/addonStore/dataManager.py [hunk 1]
```diff
@@ -162,6 +169,49 @@ def _cacheCompatibleAddons(self, addonData: str, cacheHash: Optional[str]):
...
+		if resetNewAddons == "monthly" and diffDate.days >= 30:
+			return True
+		return False
+
+	def _getResetNewAddonsDate(self) -> str:
+		resetNewAddons = config.conf["addonStore"]["resetNewAddons"]
+		if resetNewAddons == "startup":
+			# Translators: Message presented in the new add-ons tab, informing that new add-ons will be reset at startup
+			return _("Will be reset at startup")
+		lastBackupTime = os.path.getmtime(self._cacheCompatibleOldFile)
```
Comment (XLTechie) [inline source/addonStore/dataManager.py:193]: I haven't read the code, but you should be catching a file not found error here I believe.
On a new portable copy of this PR's artifact, I'm getting:

```
Traceback (most recent call last):
  File "gui\blockAction.pyc", line 79, in funcWrapper
  File "gui\__init__.pyc", line 434, in onAddonStoreCommand
  File "gui\blockAction.pyc", line 79, in funcWrapper
  File "gui\__init__.pyc", line 237, in popupSettingsDialog
  File "gui\addonStoreGui\controls\storeDialog.pyc", line 56, in __init__
  File "gui\settingsDialogs.pyc", line 225, in __init__
  File "gui\addonStoreGui\controls\storeDialog.pyc", line 94, in makeSettings
  File "gui\addonStoreGui\controls\storeDialog.pyc", line 205, in _createFilterControls
  File "gui\addonStoreGui\controls\storeDialog.pyc", line 205, in <genexpr>
  File "utils\displayString.pyc", line 62, in displayString
  File "addonStore\models\status.pyc", line 53, in _displayStringLabels
  File "addonStore\models\status.pyc", line 457, in getResetNewAddonsDate
  File "addonStore\dataManager.pyc", line 193, in _getResetNewAddonsDate
  File "<frozen genericpath>", line 55, in getmtime
FileNotFoundError: [WinError 2] The system cannot find the file specified: 'C:\\Users\\luke\\Desktop\\NVDA versions\\pr4\\userConfig\\addonStore\\_cachedCompatibleAddons-old.json'
```

File: source/gui/addonStoreGui/controls/storeDialog.py [hunk 2]
```diff
@@ -195,6 +196,19 @@ def _createFilterControls(self, filterCtrlHelper: guiHelper.BoxSizerHelper) -> N
 		self.enabledFilterCtrl.Bind(wx.EVT_CHOICE, self.onEnabledFilterChange, self.enabledFilterCtrl)
 		self.bindHelpEvent("AddonStoreFilterEnabled", self.enabledFilterCtrl)
 
+		self.newFilterCtrl = cast(
+			wx.Choice,
+			filterCtrlsLine0.addLabeledControl(
+				# Translators: The label of a selection field to filter the list of add-ons in the add-on store dialog.
+				labelText=pgettext("addonStore", "Published recently/all:"),
```
Comment (XLTechie) [inline source/gui/addonStoreGui/controls/storeDialog.py:203]:
```suggestion
				labelText=pgettext("addonStore", "Release date filter:"),
```

File: source/addonStore/dataManager.py [hunk 3]
```diff
@@ -162,6 +169,49 @@ def _cacheCompatibleAddons(self, addonData: str, cacheHash: Optional[str]):
...
+		lastBackupDate = datetime.fromtimestamp(lastBackupTime)
+		formattedLastBackupDate = lastBackupDate.strftime("%d/%m/%Y")
+		if resetNewAddons == "monthly":
+			timedeltaDays = 30
+		else:  # weekly
+			timedeltaDays = 7
+		nextResetDate = lastBackupDate + timedelta(days=timedeltaDays)
+		formattedNextResetDate = nextResetDate.strftime("%d/%m/%Y")
```
Comment (XLTechie) [inline source/addonStore/dataManager.py:201]:
```suggestion
		formattedNextResetDate = nextResetDate.strftime("%x")
```

File: source/addonStore/dataManager.py [hunk 4]
```diff
@@ -162,6 +169,49 @@ def _cacheCompatibleAddons(self, addonData: str, cacheHash: Optional[str]):
...
+		return False
+
+	def _getResetNewAddonsDate(self) -> str:
+		resetNewAddons = config.conf["addonStore"]["resetNewAddons"]
+		if resetNewAddons == "startup":
+			# Translators: Message presented in the new add-ons tab, informing that new add-ons will be reset at startup
+			return _("Will be reset at startup")
+		lastBackupTime = os.path.getmtime(self._cacheCompatibleOldFile)
+		lastBackupDate = datetime.fromtimestamp(lastBackupTime)
+		formattedLastBackupDate = lastBackupDate.strftime("%d/%m/%Y")
```
Comment (XLTechie) [inline source/addonStore/dataManager.py:195]: If it's going to use a purely numeric date, I think it should use the locale-specific date representation.

```suggestion
		formattedLastBackupDate = lastBackupDate.strftime("%x")
```

File: source/addonStore/dataManager.py [hunk 5]
```diff
@@ -162,6 +169,49 @@ def _cacheCompatibleAddons(self, addonData: str, cacheHash: Optional[str]):
...
+			return True
+		if resetNewAddons == "monthly" and diffDate.days >= 30:
+			return True
+		return False
+
+	def _getResetNewAddonsDate(self) -> str:
+		resetNewAddons = config.conf["addonStore"]["resetNewAddons"]
+		if resetNewAddons == "startup":
+			# Translators: Message presented in the new add-ons tab, informing that new add-ons will be reset at startup
+			return _("Will be reset at startup")
```
Comment (XLTechie) [inline source/addonStore/dataManager.py:192]:
```suggestion
			return _("Will be reset at NVDA startup")
```

File: source/addonStore/dataManager.py [hunk 6]
```diff
@@ -162,6 +169,49 @@ def _cacheCompatibleAddons(self, addonData: str, cacheHash: Optional[str]):
...
+		lastBackupTime = os.path.getmtime(self._cacheCompatibleOldFile)
+		lastBackupDate = datetime.fromtimestamp(lastBackupTime)
+		formattedLastBackupDate = lastBackupDate.strftime("%d/%m/%Y")
+		if resetNewAddons == "monthly":
+			timedeltaDays = 30
```
Comment (XLTechie) [inline source/addonStore/dataManager.py:197]: More months have 31 days, maybe use that?

```suggestion
			timedeltaDays = 31
```

File: source/addonStore/dataManager.py [hunk 7]
```diff
@@ -162,6 +169,49 @@ def _cacheCompatibleAddons(self, addonData: str, cacheHash: Optional[str]):
...
+		else:  # weekly
+			timedeltaDays = 7
+		nextResetDate = lastBackupDate + timedelta(days=timedeltaDays)
```
Comment (XLTechie) [inline source/addonStore/dataManager.py:200]: The way this is presenting in the UI, makes it look like it's doing the math inversely. That is, on my system, if I don't choose all, the filtering date appears to be trying to show add-ons that are new between July 4th and August 3rd. I think it should be instead between June 3rd to July 4th.

**Review by XLTechie — CHANGES_REQUESTED** [review]
Some very preliminary comments, recognizing that this is a work in progress.

File: source/addonStore/dataManager.py (hunk 7 above)
Comment (nvdaes) [reply source/addonStore/dataManager.py:203]: Hi @XLTechie , thanks for your review. No, add-ons will be considered new from 4 July, and will be reset on August. Add-on metadata doesn't have a timestamp, so NVDA will start saving all compatible add-ons cnsidering them as old add-ons, and add-ons published from 4 Jul will be considered as new add-ons. Hope this helps.

**Comment by nvdaes** [comment]
Hi @XLTechie , I think I've addressed all your comments. I'll fix the schedule job to show new add-ons at startup this weekend.

**Comment by nvdaes** [comment]
@XLTechie  (%x) is not showing dates in the local format. I think we should try to use winKernel. I'll test it this weekend.

**Comment by nvdaes** [comment]
@coderabbitai full review

File: source/addonStore/dataManager.py (hunk 5 above)
Comment (XLTechie) [reply source/addonStore/dataManager.py:192]: @nvdaes you may have missed this one, or at least it's not showing as resolved at my end.

File: source/addonStore/dataManager.py (hunk 7 above)
Comment (XLTechie) [reply source/addonStore/dataManager.py:203]: @seanbudd how hard would it be to start embedding a generation timestamp in the JSON? I should think, not very.

File: source/addonStore/dataManager.py (hunk 4 above)
Comment (XLTechie) [reply source/addonStore/dataManager.py:198]: @nvdaes if it's only ever going to show, at most, 30 days, then the year isn't even necessary. In that case, using something like `"%d %B"` is, imho, more user friendly.

That said, if you're going to use a full date, even the AI agreed you should use `"%x"`. You said it doesn't work, but I am curious how it fails, as `%x` has been standard for a long, long time.

**Review by XLTechie — CHANGES_REQUESTED** [review]

**Comment by nvdaes** [comment]
> @nvdaes if it's only ever going to show, at most, 30 days, then the year isn't even necessary. In that case, using something like "%d %B" is, imho, more user friendly.
That said, if you're going to use a full date, even the AI agreed you should use "%x". You said it doesn't work, but I am curious how it fails, as %x has been standard for a long, long time.

On my system, %x returns 8/5, and in Spanish this is interpreted as May 8, not August 5.
Also, I¡ve tried to commit all your suggestions including the use of 31 days instead of 30 for months.
But I think we should include the year, since if we start saving old add-ons on 31 December 2024, I think it's better to include the 2025 year.
I think we can use this procedure, caching old compatible add-ons, and if a timestamp is added to add-ons some day, we can use another system replacing this or as an alternative.

**Comment by nvdaes** [comment]
I'll test again with %x to show dates, since testing this in the Python console results are right.

File: source/gui/settingsDialogs.py [hunk 8]
```diff
@@ -3102,10 +3104,30 @@ def makeSettings(self, settingsSizer: wx.BoxSizer) -> None:
...
+		index = [x.value for x in ShowNewAddons].index(config.conf["addonStore"]["showNewAddons"])
+		self.showNewAddonsComboBox.SetSelection(index)
+		# Translators: This is a label for the reset new add-ons combo box in the Add-on Store Settings dialog.
+		resetNewAddonsLabelText = _("&Reset new add-ons:")
+		self.resetNewAddonsComboBox = sHelper.addLabeledControl(
+			resetNewAddonsLabelText, wx.Choice, choices=[mode.displayString for mode in ResetNewAddons]
+		)
+		self.bindHelpEvent("ResetNewAddons", self.resetNewAddonsComboBox)
+		index = [x.value for x in ResetNewAddons].index(config.conf["addonStore"]["resetNewAddons"])
+		self.resetNewAddonsComboBox.SetSelection(index)
```
[the remark this answers is not in the document]
Comment (nvdaes) [reply source/gui/settingsDialogs.py:3122]: I will ignore this comment in accordance with NVDA's core code.

File: source/addonStore/dataManager.py [hunk 9]
```diff
@@ -287,6 +340,26 @@ def getLatestAddons(
...
+		installedAddons = self._installedAddonsCache._get_installedAddons()
+		for channel in compatibleAddons:
+			for addonId in compatibleAddons[channel]:
+				compatibleAddon = compatibleAddons[channel][addonId]
+				if (
+					addonId not in oldAddons[channel]
+					or compatibleAddon.addonVersionNumber != oldAddons[channel][addonId].addonVersionNumber
+				) and addonId not in installedAddons:
+					return True
+		return False
```
[the remark this answers is not in the document]
Comment (nvdaes) [reply source/addonStore/dataManager.py:355]: I'll ignore this comment since NVDA's core code doesn't have unit tests for the addonStore.

**Comment by nvdaes** [comment]
@XLTechie, in fact %x works as expected to format dates. The problem was that I was testing from source but with NVDA set to English.
I'll remove code to notify about new add-ons at NVDA startup, since using a combo box, not a dedicated tab, imo this doesn't provide a good experience and it is not useful.

**Comment by nvdaes** [comment]
@coderabbitai resolve

**Comment by nvdaes** [comment]
@coderabbitai full review

**Comment by nvdaes** [comment]
@coderabbitai review

**Comment by nvdaes** [comment]
@coderabbitai summary

**Comment by nvdaes** [comment]
Luke wrote:

> @seanbudd how hard would it be to start embedding a generation timestamp in the JSON? 

I understand that this would require to work in several repos, but, even if the release date is not presented in a column of the add-on list, I'd like to be timestamps added for the issue related to this PR. Now,the first time we start NVDA without cached old add-ons, the list is empty, and timestamps will make this cleaner.
If needed, I'll try to work in the store or the staging repo, and add-on-datastore-validation to add timestamps.

**Comment by seanbudd** [comment]
@nvdaes - thanks for investigating this approach. Given issues around the UX, we think it would be best to go with the solution of adding timestamps and the ability to sort by them. 
Closing this PR.

**Comment by nvdaes** [comment]
OK @seanbudd . I may create prs fr timestamps in the addon-datastore repo, in the addon-datastore-validation repo and in NVDA.

**Pull request closed without merging**